# BACK MARKET: ORDINATEUR RECONDITIONNE

In [143]:
import requests as r
import pandas as pd
import json
import time

In [2]:
url='https://9x8zudunn9-dsn.algolia.net/1/indexes/*/queries?x-algolia-agent=Algolia%20for%20JavaScript%20(4.8.0)%3B%20Browser%20(lite)%3B%20JS%20Helper%20(3.3.2)&x-algolia-api-key=38f3a5eacb0bd01a763275f972706859&x-algolia-application-id=9X8ZUDUNN9'

In [3]:
headers="""Accept: */*
Accept-Encoding: gzip, deflate, br
Accept-Language: it-IT,it;q=0.9,fr-FR;q=0.8,fr;q=0.7,en-US;q=0.6,en;q=0.5
Cache-Control: no-cache
Connection: keep-alive
Content-Length: 4696
content-type: application/x-www-form-urlencoded
Host: 9x8zudunn9-dsn.algolia.net
Origin: https://www.backmarket.fr
Pragma: no-cache
Referer: https://www.backmarket.fr/
Sec-Fetch-Dest: empty
Sec-Fetch-Mode: cors
Sec-Fetch-Site: cross-site
User-Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/86.0.4240.198 Safari/537.36"""

In [4]:
headers=dict([i.split(': ') for i in headers.split('\n')])

In [5]:
data="""{"requests":[{"indexName":"prod_index_starting_from","params":"hitsPerPage=30&facetingAfterDistinct=true&attributesToRetrieve=%5B%22objectID%22%2C%22backmarketID%22%2C%22link%22%2C%22price%22%2C%22price_with_currency%22%2C%22price_new%22%2C%22price_new_with_currency%22%2C%22warranty%22%2C%22title%22%2C%22slug%22%2C%22merchant_slug%22%2C%22sub_title%22%2C%22state_label%22%2C%22state_value%22%2C%22image1%22%2C%22compatible_carriers%22%2C%22is_sim_locked%22%2C%22list_view%22%2C%22nb_results_starting_from%22%2C%22stockRaw%22%2C%22pressureStockLimit%22%5D&page=1&facets=%5B%22category_3%22%2C%22screen_size%22%2C%22brand%22%2C%22processor_type%22%2C%22keyboard_type%22%2C%22memory%22%2C%22os%22%2C%22storage_type%22%2C%22states_list%22%2C%22model%22%2C%22storage%22%2C%22graphic_card%22%2C%22payment_methods%22%2C%22warranty_with_unit%22%2C%22shipping_delay%22%2C%22shipping_free%22%2C%22price_ranges.sm-1%22%2C%22price_ranges.sm-2%22%2C%22price_ranges.md-1%22%2C%22price_ranges.md-1b%22%2C%22price_ranges.md-1c%22%2C%22price_ranges.md-2%22%2C%22price_ranges.lg-1%22%2C%22price_ranges.lg-2%22%2C%22price_ranges.lg-3%22%2C%22price%22%2C%22cat_id%22%5D&tagFilters=&facetFilters=%5B%5B%22cat_id%3A7%22%2C%22cat_id%3A9%22%2C%22cat_id%3A11%22%2C%22cat_id%3A12%22%2C%22cat_id%3A109%22%2C%22cat_id%3A110%22%2C%22cat_id%3A111%22%2C%22cat_id%3A112%22%2C%22cat_id%3A113%22%2C%22cat_id%3A114%22%2C%22cat_id%3A115%22%2C%22cat_id%3A116%22%2C%22cat_id%3A117%22%5D%5D&numericFilters=%5B%5B%22buybox%3D2%22%5D%2C%5B%22publication_state%3D2%22%5D%2C%5B%22special_offer_type%3D0%22%5D%5D"},{"indexName":"prod_index_starting_from","params":"hitsPerPage=1&facetingAfterDistinct=true&attributesToRetrieve=%5B%22objectID%22%2C%22backmarketID%22%2C%22link%22%2C%22price%22%2C%22price_with_currency%22%2C%22price_new%22%2C%22price_new_with_currency%22%2C%22warranty%22%2C%22title%22%2C%22slug%22%2C%22merchant_slug%22%2C%22sub_title%22%2C%22state_label%22%2C%22state_value%22%2C%22image1%22%2C%22compatible_carriers%22%2C%22is_sim_locked%22%2C%22list_view%22%2C%22nb_results_starting_from%22%2C%22stockRaw%22%2C%22pressureStockLimit%22%5D&page=0&attributesToHighlight=%5B%5D&attributesToSnippet=%5B%5D&tagFilters=&analytics=false&clickAnalytics=false&facets=cat_id&numericFilters=%5B%5B%22buybox%3D2%22%5D%2C%5B%22publication_state%3D2%22%5D%2C%5B%22special_offer_type%3D0%22%5D%5D"},{"indexName":"prod_index_starting_from","params":"hitsPerPage=30&facetingAfterDistinct=true&attributesToRetrieve=%5B%22objectID%22%2C%22backmarketID%22%2C%22link%22%2C%22price%22%2C%22price_with_currency%22%2C%22price_new%22%2C%22price_new_with_currency%22%2C%22warranty%22%2C%22title%22%2C%22slug%22%2C%22merchant_slug%22%2C%22sub_title%22%2C%22state_label%22%2C%22state_value%22%2C%22image1%22%2C%22compatible_carriers%22%2C%22is_sim_locked%22%2C%22list_view%22%2C%22nb_results_starting_from%22%2C%22stockRaw%22%2C%22pressureStockLimit%22%5D&page=1&facets=%5B%22category_3%22%2C%22screen_size%22%2C%22brand%22%2C%22processor_type%22%2C%22keyboard_type%22%2C%22memory%22%2C%22os%22%2C%22storage_type%22%2C%22states_list%22%2C%22model%22%2C%22storage%22%2C%22graphic_card%22%2C%22payment_methods%22%2C%22warranty_with_unit%22%2C%22shipping_delay%22%2C%22shipping_free%22%2C%22price_ranges.sm-1%22%2C%22price_ranges.sm-2%22%2C%22price_ranges.md-1%22%2C%22price_ranges.md-1b%22%2C%22price_ranges.md-1c%22%2C%22price_ranges.md-2%22%2C%22price_ranges.lg-1%22%2C%22price_ranges.lg-2%22%2C%22price_ranges.lg-3%22%2C%22cat_id%22%5D&tagFilters=&facetFilters=%5B%5B%22cat_id%3A7%22%2C%22cat_id%3A9%22%2C%22cat_id%3A11%22%2C%22cat_id%3A12%22%2C%22cat_id%3A109%22%2C%22cat_id%3A110%22%2C%22cat_id%3A111%22%2C%22cat_id%3A112%22%2C%22cat_id%3A113%22%2C%22cat_id%3A114%22%2C%22cat_id%3A115%22%2C%22cat_id%3A116%22%2C%22cat_id%3A117%22%5D%5D&numericFilters=%5B%5B%22buybox%3D2%22%5D%2C%5B%22publication_state%3D2%22%5D%2C%5B%22special_offer_type%3D0%22%5D%5D"},{"indexName":"prod_index_starting_from","params":"hitsPerPage=1&facetingAfterDistinct=true&attributesToRetrieve=%5B%22objectID%22%2C%22backmarketID%22%2C%22link%22%2C%22price%22%2C%22price_with_currency%22%2C%22price_new%22%2C%22price_new_with_currency%22%2C%22warranty%22%2C%22title%22%2C%22slug%22%2C%22merchant_slug%22%2C%22sub_title%22%2C%22state_label%22%2C%22state_value%22%2C%22image1%22%2C%22compatible_carriers%22%2C%22is_sim_locked%22%2C%22list_view%22%2C%22nb_results_starting_from%22%2C%22stockRaw%22%2C%22pressureStockLimit%22%5D&page=0&attributesToHighlight=%5B%5D&attributesToSnippet=%5B%5D&tagFilters=&analytics=false&clickAnalytics=false&facets=cat_id&numericFilters=%5B%5B%22buybox%3D2%22%5D%2C%5B%22publication_state%3D2%22%5D%2C%5B%22special_offer_type%3D0%22%5D%5D"}]}"""

In [6]:
response=r.post(url,headers=headers, data=data)

In [ ]:
back=response.json() # convert the response from code to json file
back

In [37]:
# back.keys()

In [38]:
# back['results'][0]['hits'][0]['link']['params']['slug']

In [39]:
# df=pd.json_normalize(back['results'][0]['hits'])


In [9]:
headers="""Accept: */*
Accept-Encoding: gzip, deflate, br
Accept-Language: it-IT,it;q=0.9,fr-FR;q=0.8,fr;q=0.7,en-US;q=0.6,en;q=0.5
Cache-Control: no-cache
Connection: keep-alive
Content-Length: 4696
content-type: application/x-www-form-urlencoded
Host: 9x8zudunn9-dsn.algolia.net
Origin: https://www.backmarket.fr
Pragma: no-cache
Referer: https://www.backmarket.fr/
Sec-Fetch-Dest: empty
Sec-Fetch-Mode: cors
Sec-Fetch-Site: cross-site
User-Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/86.0.4240.198 Safari/537.36"""

headers=dict([i.split(': ') for i in headers.split('\n')]) 


In [10]:
data='''{"requests":[{"indexName":"prod_index_starting_from","params":"hitsPerPage=30&facetingAfterDistinct=true&attributesToRetrieve=%5B%22objectID%22%2C%22backmarketID%22%2C%22link%22%2C%22price%22%2C%22price_with_currency%22%2C%22price_new%22%2C%22price_new_with_currency%22%2C%22warranty%22%2C%22title%22%2C%22slug%22%2C%22merchant_slug%22%2C%22sub_title%22%2C%22state_label%22%2C%22state_value%22%2C%22image1%22%2C%22compatible_carriers%22%2C%22is_sim_locked%22%2C%22list_view%22%2C%22nb_results_starting_from%22%2C%22stockRaw%22%2C%22pressureStockLimit%22%5D&page=1&facets=%5B%22category_3%22%2C%22screen_size%22%2C%22brand%22%2C%22processor_type%22%2C%22keyboard_type%22%2C%22memory%22%2C%22os%22%2C%22storage_type%22%2C%22states_list%22%2C%22model%22%2C%22storage%22%2C%22graphic_card%22%2C%22payment_methods%22%2C%22warranty_with_unit%22%2C%22shipping_delay%22%2C%22shipping_free%22%2C%22price_ranges.sm-1%22%2C%22price_ranges.sm-2%22%2C%22price_ranges.md-1%22%2C%22price_ranges.md-1b%22%2C%22price_ranges.md-1c%22%2C%22price_ranges.md-2%22%2C%22price_ranges.lg-1%22%2C%22price_ranges.lg-2%22%2C%22price_ranges.lg-3%22%2C%22price%22%2C%22cat_id%22%5D&tagFilters=&facetFilters=%5B%5B%22cat_id%3A7%22%2C%22cat_id%3A9%22%2C%22cat_id%3A11%22%2C%22cat_id%3A12%22%2C%22cat_id%3A109%22%2C%22cat_id%3A110%22%2C%22cat_id%3A111%22%2C%22cat_id%3A112%22%2C%22cat_id%3A113%22%2C%22cat_id%3A114%22%2C%22cat_id%3A115%22%2C%22cat_id%3A116%22%2C%22cat_id%3A117%22%5D%5D&numericFilters=%5B%5B%22buybox%3D2%22%5D%2C%5B%22publication_state%3D2%22%5D%2C%5B%22special_offer_type%3D0%22%5D%5D"},{"indexName":"prod_index_starting_from","params":"hitsPerPage=1&facetingAfterDistinct=true&attributesToRetrieve=%5B%22objectID%22%2C%22backmarketID%22%2C%22link%22%2C%22price%22%2C%22price_with_currency%22%2C%22price_new%22%2C%22price_new_with_currency%22%2C%22warranty%22%2C%22title%22%2C%22slug%22%2C%22merchant_slug%22%2C%22sub_title%22%2C%22state_label%22%2C%22state_value%22%2C%22image1%22%2C%22compatible_carriers%22%2C%22is_sim_locked%22%2C%22list_view%22%2C%22nb_results_starting_from%22%2C%22stockRaw%22%2C%22pressureStockLimit%22%5D&page=0&attributesToHighlight=%5B%5D&attributesToSnippet=%5B%5D&tagFilters=&analytics=false&clickAnalytics=false&facets=cat_id&numericFilters=%5B%5B%22buybox%3D2%22%5D%2C%5B%22publication_state%3D2%22%5D%2C%5B%22special_offer_type%3D0%22%5D%5D"},{"indexName":"prod_index_starting_from","params":"hitsPerPage=30&facetingAfterDistinct=true&attributesToRetrieve=%5B%22objectID%22%2C%22backmarketID%22%2C%22link%22%2C%22price%22%2C%22price_with_currency%22%2C%22price_new%22%2C%22price_new_with_currency%22%2C%22warranty%22%2C%22title%22%2C%22slug%22%2C%22merchant_slug%22%2C%22sub_title%22%2C%22state_label%22%2C%22state_value%22%2C%22image1%22%2C%22compatible_carriers%22%2C%22is_sim_locked%22%2C%22list_view%22%2C%22nb_results_starting_from%22%2C%22stockRaw%22%2C%22pressureStockLimit%22%5D&page=1&facets=%5B%22category_3%22%2C%22screen_size%22%2C%22brand%22%2C%22processor_type%22%2C%22keyboard_type%22%2C%22memory%22%2C%22os%22%2C%22storage_type%22%2C%22states_list%22%2C%22model%22%2C%22storage%22%2C%22graphic_card%22%2C%22payment_methods%22%2C%22warranty_with_unit%22%2C%22shipping_delay%22%2C%22shipping_free%22%2C%22price_ranges.sm-1%22%2C%22price_ranges.sm-2%22%2C%22price_ranges.md-1%22%2C%22price_ranges.md-1b%22%2C%22price_ranges.md-1c%22%2C%22price_ranges.md-2%22%2C%22price_ranges.lg-1%22%2C%22price_ranges.lg-2%22%2C%22price_ranges.lg-3%22%2C%22cat_id%22%5D&tagFilters=&facetFilters=%5B%5B%22cat_id%3A7%22%2C%22cat_id%3A9%22%2C%22cat_id%3A11%22%2C%22cat_id%3A12%22%2C%22cat_id%3A109%22%2C%22cat_id%3A110%22%2C%22cat_id%3A111%22%2C%22cat_id%3A112%22%2C%22cat_id%3A113%22%2C%22cat_id%3A114%22%2C%22cat_id%3A115%22%2C%22cat_id%3A116%22%2C%22cat_id%3A117%22%5D%5D&numericFilters=%5B%5B%22buybox%3D2%22%5D%2C%5B%22publication_state%3D2%22%5D%2C%5B%22special_offer_type%3D0%22%5D%5D"},{"indexName":"prod_index_starting_from","params":"hitsPerPage=1&facetingAfterDistinct=true&attributesToRetrieve=%5B%22objectID%22%2C%22backmarketID%22%2C%22link%22%2C%22price%22%2C%22price_with_currency%22%2C%22price_new%22%2C%22price_new_with_currency%22%2C%22warranty%22%2C%22title%22%2C%22slug%22%2C%22merchant_slug%22%2C%22sub_title%22%2C%22state_label%22%2C%22state_value%22%2C%22image1%22%2C%22compatible_carriers%22%2C%22is_sim_locked%22%2C%22list_view%22%2C%22nb_results_starting_from%22%2C%22stockRaw%22%2C%22pressureStockLimit%22%5D&page=0&attributesToHighlight=%5B%5D&attributesToSnippet=%5B%5D&tagFilters=&analytics=false&clickAnalytics=false&facets=cat_id&numericFilters=%5B%5B%22buybox%3D2%22%5D%2C%5B%22publication_state%3D2%22%5D%2C%5B%22special_offer_type%3D0%22%5D%5D"}]}'''

In [107]:
df_final=pd.DataFrame()
for i in range(1,10):
    url='https://9x8zudunn9-dsn.algolia.net/1/indexes/*/queries?x-algolia-agent=Algolia%20for%20JavaScript%20(4.8.0)%3B%20Browser%20(lite)%3B%20JS%20Helper%20(3.3.2)&x-algolia-api-key=38f3a5eacb0bd01a763275f972706859&x-algolia-application-id=9X8ZUDUNN9'
    data='{"requests":[{"indexName":"prod_index_starting_from","params":"hitsPerPage=30&facetingAfterDistinct=true&attributesToRetrieve=%5B%22objectID%22%2C%22backmarketID%22%2C%22link%22%2C%22price%22%2C%22price_with_currency%22%2C%22price_new%22%2C%22price_new_with_currency%22%2C%22warranty%22%2C%22title%22%2C%22slug%22%2C%22merchant_slug%22%2C%22sub_title%22%2C%22state_label%22%2C%22state_value%22%2C%22image1%22%2C%22compatible_carriers%22%2C%22is_sim_locked%22%2C%22list_view%22%2C%22nb_results_starting_from%22%2C%22stockRaw%22%2C%22pressureStockLimit%22%5D&page=0&facets=%5B%22category_3%22%2C%22screen_size%22%2C%22brand%22%2C%22processor_type%22%2C%22keyboard_type%22%2C%22memory%22%2C%22os%22%2C%22storage_type%22%2C%22states_list%22%2C%22model%22%2C%22storage%22%2C%22graphic_card%22%2C%22payment_methods%22%2C%22warranty_with_unit%22%2C%22shipping_delay%22%2C%22shipping_free%22%2C%22price_ranges.sm-1%22%2C%22price_ranges.sm-2%22%2C%22price_ranges.md-1%22%2C%22price_ranges.md-1b%22%2C%22price_ranges.md-1c%22%2C%22price_ranges.md-2%22%2C%22price_ranges.lg-1%22%2C%22price_ranges.lg-2%22%2C%22price_ranges.lg-3%22%2C%22price%22%2C%22cat_id%22%5D&tagFilters=&facetFilters=%5B%5B%22cat_id%3A7%22%2C%22cat_id%3A9%22%2C%22cat_id%3A11%22%2C%22cat_id%3A12%22%2C%22cat_id%3A109%22%2C%22cat_id%3A110%22%2C%22cat_id%3A111%22%2C%22cat_id%3A112%22%2C%22cat_id%3A113%22%2C%22cat_id%3A114%22%2C%22cat_id%3A115%22%2C%22cat_id%3A116%22%2C%22cat_id%3A117%22%5D%5D&numericFilters=%5B%5B%22buybox%3D2%22%5D%2C%5B%22publication_state%3D2%22%5D%2C%5B%22special_offer_type%3D0%22%5D%5D"},{"indexName":"prod_index_starting_from","params":"hitsPerPage=1&facetingAfterDistinct=true&attributesToRetrieve=%5B%22objectID%22%2C%22backmarketID%22%2C%22link%22%2C%22price%22%2C%22price_with_currency%22%2C%22price_new%22%2C%22price_new_with_currency%22%2C%22warranty%22%2C%22title%22%2C%22slug%22%2C%22merchant_slug%22%2C%22sub_title%22%2C%22state_label%22%2C%22state_value%22%2C%22image1%22%2C%22compatible_carriers%22%2C%22is_sim_locked%22%2C%22list_view%22%2C%22nb_results_starting_from%22%2C%22stockRaw%22%2C%22pressureStockLimit%22%5D&page=0&attributesToHighlight=%5B%5D&attributesToSnippet=%5B%5D&tagFilters=&analytics=false&clickAnalytics=false&facets=cat_id&numericFilters=%5B%5B%22buybox%3D2%22%5D%2C%5B%22publication_state%3D2%22%5D%2C%5B%22special_offer_type%3D0%22%5D%5D"},{"indexName":"prod_index_starting_from","params":"hitsPerPage=30&facetingAfterDistinct=true&attributesToRetrieve=%5B%22objectID%22%2C%22backmarketID%22%2C%22link%22%2C%22price%22%2C%22price_with_currency%22%2C%22price_new%22%2C%22price_new_with_currency%22%2C%22warranty%22%2C%22title%22%2C%22slug%22%2C%22merchant_slug%22%2C%22sub_title%22%2C%22state_label%22%2C%22state_value%22%2C%22image1%22%2C%22compatible_carriers%22%2C%22is_sim_locked%22%2C%22list_view%22%2C%22nb_results_starting_from%22%2C%22stockRaw%22%2C%22pressureStockLimit%22%5D&page=1&facets=%5B%22category_3%22%2C%22screen_size%22%2C%22brand%22%2C%22processor_type%22%2C%22keyboard_type%22%2C%22memory%22%2C%22os%22%2C%22storage_type%22%2C%22states_list%22%2C%22model%22%2C%22storage%22%2C%22graphic_card%22%2C%22payment_methods%22%2C%22warranty_with_unit%22%2C%22shipping_delay%22%2C%22shipping_free%22%2C%22price_ranges.sm-1%22%2C%22price_ranges.sm-2%22%2C%22price_ranges.md-1%22%2C%22price_ranges.md-1b%22%2C%22price_ranges.md-1c%22%2C%22price_ranges.md-2%22%2C%22price_ranges.lg-1%22%2C%22price_ranges.lg-2%22%2C%22price_ranges.lg-3%22%2C%22cat_id%22%5D&tagFilters=&facetFilters=%5B%5B%22cat_id%3A7%22%2C%22cat_id%3A9%22%2C%22cat_id%3A11%22%2C%22cat_id%3A12%22%2C%22cat_id%3A109%22%2C%22cat_id%3A110%22%2C%22cat_id%3A111%22%2C%22cat_id%3A112%22%2C%22cat_id%3A113%22%2C%22cat_id%3A114%22%2C%22cat_id%3A115%22%2C%22cat_id%3A116%22%2C%22cat_id%3A117%22%5D%5D&numericFilters=%5B%5B%22buybox%3D2%22%5D%2C%5B%22publication_state%3D2%22%5D%2C%5B%22special_offer_type%3D0%22%5D%5D"},{"indexName":"prod_index_starting_from","params":"hitsPerPage=1&facetingAfterDistinct=true&attributesToRetrieve=%5B%22objectID%22%2C%22backmarketID%22%2C%22link%22%2C%22price%22%2C%22price_with_currency%22%2C%22price_new%22%2C%22price_new_with_currency%22%2C%22warranty%22%2C%22title%22%2C%22slug%22%2C%22merchant_slug%22%2C%22sub_title%22%2C%22state_label%22%2C%22state_value%22%2C%22image1%22%2C%22compatible_carriers%22%2C%22is_sim_locked%22%2C%22list_view%22%2C%22nb_results_starting_from%22%2C%22stockRaw%22%2C%22pressureStockLimit%22%5D&page=0&attributesToHighlight=%5B%5D&attributesToSnippet=%5B%5D&tagFilters=&analytics=false&clickAnalytics=false&facets=cat_id&numericFilters=%5B%5B%22buybox%3D2%22%5D%2C%5B%22publication_state%3D2%22%5D%2C%5B%22special_offer_type%3D0%22%5D%5D"}]}'
    data = data.replace("page=0", f"page={i}")
    response=r.post(url, headers=headers, data=data)
    back=response.json()
    df=pd.json_normalize(back['results'][0]['hits'])
    df_final=df_final.append(df)
    print('page number',i,'is done!')
    if i%5==0:
        time.sleep(2)

page number 1 is done!
page number 2 is done!
page number 3 is done!
page number 4 is done!
page number 5 is done!
page number 6 is done!
page number 7 is done!
page number 8 is done!
page number 9 is done!


In [117]:
#type(response.json())

dict

In [112]:
#response.text.find("paginationLimitedTo")>1

True

# Data Cleaning

In [ ]:
df_final.info()

In [109]:
df_final.shape

(270, 43)

In [110]:
df.columns

Index(['backmarketID', 'price_with_currency', 'price', 'stockRaw',
       'pressureStockLimit', 'warranty', 'slug', 'state_value', 'state_label',
       'compatible_carriers', 'is_sim_locked', 'title', 'price_new',
       'price_new_with_currency', 'merchant_slug', 'image1', 'objectID',
       'link.type', 'link.name', 'link.params.id', 'link.params.slug',
       'link.hash.l', 'list_view.brand', 'list_view.warranty',
       'list_view.screen_size', 'list_view.storage', 'list_view.storage_type',
       'list_view.memory', 'list_view.model', 'list_view.processor_speed',
       'list_view.processor_type', 'list_view.date_release',
       'list_view.keyboard_type_language', '_highlightResult.title.value',
       '_highlightResult.title.matchLevel',
       '_highlightResult.title.matchedWords', '_highlightResult.brand.value',
       '_highlightResult.brand.matchLevel',
       '_highlightResult.brand.matchedWords',
       '_highlightResult.category_3.value',
       '_highlightResult.categor

In [ ]:
df_final.info()

Conclusion:
    - 270 rows vs 43 columns. Enough data, probably need to drop some columns
    - Lots of columns start with the same prefix or contained weird characters 
    - Data type: boolean, float, integer and object (>75%)
    - info(): shows us that there is only one column with all missing data (list_view)

## Drop Columns

In [112]:
df_final.drop(columns=['_highlightResult.title.matchLevel','compatible_carriers','image1', 'link.type', 'link.name', 'link.params.id', 'link.params.slug',
'link.hash.l','_highlightResult.title.matchLevel','_highlightResult.title.matchedWords', '_highlightResult.brand.value',
       '_highlightResult.brand.matchLevel',
       '_highlightResult.brand.matchedWords',
       '_highlightResult.category_3.value',
       '_highlightResult.category_3.matchLevel',
       '_highlightResult.category_3.matchedWords', 'list_view','is_sim_locked','price_new_with_currency','price_with_currency'],inplace=True)                          

                          

#  String Operations

In [136]:
df_final.columns=df_final.columns=df_final.columns.str.lower().str.strip('_ ').str.replace('list_view.','').str.replace(".","_")    #lower case all the string
df_final.columns

Index(['backmarketid', 'price', 'stockraw', 'pressurestocklimit', 'warranty',
       'slug', 'state_value', 'state_label', 'title', 'price_new',
       'merchant_slug', 'objectid', 'brand', 'warranty', 'size', 'storage',
       'memory', 'model', 'processor_speed', 'processor_type', 'storage_type',
       'date_release', 'keyboard_type_language',
       'highlightresult_title_value'],
      dtype='object')

In [ ]:
df_final.head()

Summary of the cleaning:
    - Drop columns with all null values, with 1 unique values on the rows... not relevant to the dataset
    - String Operations: remove _ at the beginning of the column name, shorten column name etc.

# Export file

In [144]:
# Export comma-separated variable file
df_final.to_csv('df_final.csv', index=False)